fix(sqllab_export): manually encode CSV output to support utf-8-sig

Habeeb556 opened the pull request:
Manually encodes the CSV string before returning it in the response. Since `Werkzeug 3` removed support for automatic encoding.

<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

By explicitly encoding the CSV output as `utf-8-sig`, this change ensures proper rendering of non-English characters (e.g., Arabic) in tools like Excel. This is necessary because the CSV file itself isn't interpreted correctly without a BOM, and `Werkzeug 3` no longer handles encoding automatically—so this PR applies the encoding manually to maintain compatibility.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

Before:

<img width="194" height="89" alt="image" src="https://github.com/user-attachments/assets/ba69dd12-7481-4f75-b954-8724a988e85a" />

After:

<img width="199" height="71" alt="image" src="https://github.com/user-attachments/assets/61504a96-1514-4d80-817c-fde0ef50d62f" />

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [X] Has associated issue: #5372, #4506, #29224, #29410, #29506
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Habeeb556 committed:
fix(sqllab_export): manually encode CSV output to support utf-8-sig

Habeeb556 commented:
> Looks like some unit tests are failing... Not sure this won't break anything for anyone out there.

Due to a **type mismatch**, the test is comparing a **byte string** (`b'foo\n1\n2\n'`) with a **regular string** (`'foo\n1\n2\n'`), which will always fail in Python.

The test assumes the function returns a string, which is fine for displaying data. However, since the actual export returns **bytes** (raw binary data)—as expected for file downloads—the test should be updated to expect bytes instead.

Additionally, as noted in Werkzeug issue https://github.com/pallets/werkzeug/issues/3057, **automatic encoding was disabled in Werkzeug 3**, so this now needs to be handled manually. That’s what I have implemented here.

While downgrading to `Werkzeug==2.3.8` does resolve the issue temporarily, it's not ideal. That version was flagged in issue #28623 due to a **CVE**, so relying on it long-term would be risky.

mistercrunch commented:
in 2025 utf8 should be supported across the ecosystem really. So about the unit test you should be ok to cast to byte string on the right side of the assertion and pass the test (?)

Just hoping that nothing is going to break here. Someone somewhere might use an old version of Excel and report an issue at some point. Unclear whether the whole ecosystem is ready for utf8 csv (?) Wondering if we need a checkbox "export as utf8" that people can check/uncheck. Though we could move to utf8 and wait issues to be reported to decide whether we should add a checkbox?

mistercrunch commented:
My take: assume the ecosystem supports utf8 and merge this PR (after unit tests are fixed) . If/when someone reports issues with utf8 support in antiquated tools, we recommend to upgrade their Excel 95/XP to something from this era.

Habeeb556 committed:
Merge branch 'apache:master' into master-1

Habeeb556 committed:
fix(sqllab_export): manually encode CSV output to support utf-8-sig

Habeeb556 commented:
I may have miscommunicated the intent in the PR summary—my concern isn’t about outdated Excel versions, but rather the fact that the CSV file itself isn’t interpreted correctly when it contains non-English characters (like Arabic) even it's encoded with `utf-8-sig`. Since `Werkzeug 3` removed automatic encoding, this needs to be just handled manually from the hard code.

By default, the encoding is set to `utf-8`, but it can be overridden by specifying a custom value in `superset/config.py`

`CSV_EXPORT = {"encoding": "utf-8-sig"}` — this keeps the behavior configurable based on deployment needs.

Tests are now updated to match the byte output.

mistercrunch commented:
Should we deprecate `CSV_EXPORT = {"encoding": "utf-8-sig"}` then?

mistercrunch commented:
My take here is that we should just make things work properly with UTF-8 out of the box, but I get why we need the BOM now.

From my understanding, the issue is Excel - without the BOM signature bytes, Excel can't figure out that our UTF-8 CSV files are actually UTF-8, so it mangles any non-English characters. That's why `utf-8-sig` matters - it adds those 3 magic bytes (`\xEF\xBB\xBF`) that tell Excel "hey, this is UTF-8!"

But I still think we should avoid config complexity. A few thoughts:

- Let's just default to `utf-8-sig` for all CSV exports - it's backward compatible and fixes the Excel issue
- The BOM is valid UTF-8 and most modern tools handle it fine (unlike the bad old days)
- No need for a `CSV_EXPORT` config - this should "just work" without users having to know about BOMs

Can we simplify this to:
1. Always use `utf-8-sig` for CSV exports (handles Excel + international chars)  
2. Remove the config option entirely
3. Add tests for various UTF-8 scenarios (Arabic, Chinese, etc.)

What do you think about just making this the default behavior? (while noting the change in UPDATING.md)

Habeeb556 committed:
Merge branch 'apache:master' into master-1

Habeeb556 committed:
Merge branch 'apache:master' into master-1

Habeeb556 committed:
add note UPDATING.md

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
fix(sqllab_export): CSV exports now use utf-8-sig encoding by default to include a UTF-8 BOM

Habeeb556 committed:
Skip mypy for now

Habeeb556 committed:
Update api_tests.py

Habeeb556 committed:
Apply ruff auto-fixes

Habeeb556 committed:
Apply ruff-format auto-fixes

Habeeb556 committed:
fix: Add tests for various UTF-8 scenarios

Habeeb556 committed:
fix: Add tests for various UTF-8 scenarios

Habeeb556 committed:
Apply ruff-format auto-fixes

Habeeb556 committed:
Merge branch 'apache:master' into master-1

Habeeb556 committed:
fix: Remove BOM from first cell for proper comparison

Habeeb556 committed:
Apply ruff auto-fixes

Habeeb556 commented:
@mistercrunch Verified that the CSV export now correctly includes a UTF-8 BOM (`utf-8-sig`) and handles multilingual data as expected.
- Added a test case to check for BOM presence and correct parsing of Arabic and Chinese characters.
- Also updated `UPDATING.md` to reflect this change.

Everything is working as intended — ready for merge.

mistercrunch reviewed (approved):
🔥 🚀
